File: backend/services/stigmer-server/pkg/domain/agentexecution/temporal/dispatch.go
package temporal

import (
	"context"
	"errors"
	"fmt"

	"github.com/rs/zerolog/log"
	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store"
)

// DefaultActivityTaskQueue is the fallback Temporal task queue for activity
// routing. Used in global routing mode and as a fallback when session ID is
// empty in per-session mode.
const DefaultActivityTaskQueue = "stigmer_runner"

// sessionTaskQueuePrefix is the prefix for per-session task queue names.
const sessionTaskQueuePrefix = "session:"

// DispatchResult holds the resolved Temporal task queue, session harness,
// and execution target for workflow creation.
type DispatchResult struct {
	TaskQueue       string
	Harness         sessionv1.Harness
	ExecutionTarget sessionv1.ExecutionTarget
}

// FormatSessionTaskQueue derives the canonical Temporal task queue name for a
// given session ID. The format is "session:{session_id}".
//
// This is a pure function with no side effects — it can be used by any
// component that needs to derive or validate a session task queue name
// (e.g., runner boot scripts, integration tests, cloud sandbox provisioning).
func FormatSessionTaskQueue(sessionID string) string {
	return sessionTaskQueuePrefix + sessionID
}

// ResolveActivityTaskQueue determines which Temporal task queue an execution's
// activities should be routed to.
//
// When activityTaskQueueOverride is non-empty, it is used directly — this
// enables sandbox sharing where a parent workflow execution passes its own
// queue (wfexec:{id}) so child agents run in the same sandbox without
// provisioning new VMs. ExecutionTarget is set to LOCAL in this case to
// prevent EnsureSessionSandboxStep from triggering.
//
// Routing modes (controlled by Config.ActivityRouting):
//   - "global": Always returns DefaultActivityTaskQueue. All sessions share
//     one runner pool. This is the default for OSS local development.
//   - "session": Returns session:{session_id} when a session ID is available.
//     Each session routes to a dedicated runner. Used by desktop (embedded
//     runners) and cloud (per-session sandboxes).
//
// In both modes, the session is loaded to extract the harness configuration
// (NATIVE vs CURSOR) and the execution target (LOCAL vs CLOUD) which determines
// which activity type the workflow invokes and who provides the runner.
func ResolveActivityTaskQueue(ctx context.Context, s store.Store, sessionID string, cfg *Config, activityTaskQueueOverride string) (DispatchResult, error) {
	// Sandbox affinity: parent workflow already has a sandbox on this queue.
	// Route there directly — no session routing, no sandbox provisioning needed.
	if activityTaskQueueOverride != "" {
		harness := sessionv1.Harness_HARNESS_NATIVE
		if sessionID != "" {
			session := &sessionv1.Session{}
			if err := s.GetResource(ctx, apiresourcekind.ApiResourceKind_session, sessionID, session); err == nil {
				harness = session.GetSpec().GetHarness()
			}
		}

		log.Info().
			Str("session_id", sessionID).
			Str("task_queue", activityTaskQueueOverride).
			Str("override_source", "parent_workflow_sandbox").
			Msg("Dispatch using activity_task_queue override (sandbox affinity)")

		return DispatchResult{
			TaskQueue:       activityTaskQueueOverride,
			Harness:         harness,
			ExecutionTarget: sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL,
		}, nil
	}

	harness := sessionv1.Harness_HARNESS_NATIVE
	executionTarget := sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED

	if sessionID != "" {
		session := &sessionv1.Session{}
		if err := s.GetResource(ctx, apiresourcekind.ApiResourceKind_session, sessionID, session); err != nil {
			if !errors.Is(err, store.ErrNotFound) {
				return DispatchResult{}, fmt.Errorf("failed to load session for dispatch: %w", err)
			}
			log.Warn().Str("session_id", sessionID).
				Msg("Session not found during dispatch, using defaults")
		} else {
			harness = session.GetSpec().GetHarness()
			executionTarget = session.GetSpec().GetExecutionTarget()
		}
	}

	resolvedTarget := resolveExecutionTarget(executionTarget, cfg)
	taskQueue := resolveTaskQueue(sessionID, cfg)

	log.Info().
		Str("session_id", sessionID).
		Str("task_queue", taskQueue).
		Str("routing_mode", cfg.ActivityRouting).
		Str("execution_target", resolvedTarget.String()).
		Msg("Dispatch resolved activity task queue")

	return DispatchResult{
		TaskQueue:       taskQueue,
		Harness:         harness,
		ExecutionTarget: resolvedTarget,
	}, nil
}

// resolveExecutionTarget resolves UNSPECIFIED to the configured default.
func resolveExecutionTarget(target sessionv1.ExecutionTarget, cfg *Config) sessionv1.ExecutionTarget {
	if target != sessionv1.ExecutionTarget_EXECUTION_TARGET_UNSPECIFIED {
		return target
	}
	if cfg.DefaultExecutionTarget == DefaultExecutionTargetCloud {
		return sessionv1.ExecutionTarget_EXECUTION_TARGET_CLOUD
	}
	return sessionv1.ExecutionTarget_EXECUTION_TARGET_LOCAL
}

// resolveTaskQueue derives the task queue name based on routing mode and session ID.
func resolveTaskQueue(sessionID string, cfg *Config) string {
	if cfg.ActivityRouting == RoutingSession && sessionID != "" {
		return FormatSessionTaskQueue(sessionID)
	}
	return cfg.RunnerQueue
}

File: backend/services/stigmer-server/src/temporal/agentexecution/dispatch.ts
/**
 * Activity-queue dispatch resolution — ports
 * pkg/domain/agentexecution/temporal/dispatch.go.
 *
 * Decides which Temporal task queue an execution's runner activities route
 * to, and resolves the session's harness (NATIVE vs CURSOR — which
 * activity type the workflow invokes) and execution target (LOCAL vs
 * CLOUD — cloud uses it for sandbox provisioning; the OSS workflow
 * ignores it).
 *
 * The `wfexec:{id}` override lane: a parent workflow execution passes its
 * OWN queue so child agents share its sandbox. This domain honors the
 * override as an opaque string (the prefix constant belongs to
 * workflowexecution, #21) and forces ExecutionTarget LOCAL to suppress
 * sandbox provisioning — exactly Go's contract.
 *
 * Proven by the agentexecution suites on local-ts-execution and by the
 * co-located dispatch tests (ports dispatch_test.go).
 */
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";
import { SessionSchema } from "@stigmer/protos/ai/stigmer/agentic/session/v1/api_pb";
import {
  ExecutionTarget,
  Harness,
} from "@stigmer/protos/ai/stigmer/agentic/session/v1/enum_pb";

import type { Logger } from "../../boot/logger.js";
import {
  AgentExecutionTemporalConfig,
  ROUTING_SESSION,
} from "../../domain/agentexecution/temporal/config.js";
import { ResourceNotFoundError, type Store } from "../../store/interface.js";

/** Go dispatch.go sessionTaskQueuePrefix. */
const SESSION_TASK_QUEUE_PREFIX = "session:";

/** Go temporal.DispatchResult. */
export interface DispatchResult {
  readonly taskQueue: string;
  readonly harness: Harness;
  readonly executionTarget: ExecutionTarget;
}

/**
 * Derives the canonical per-session task queue name (Go
 * FormatSessionTaskQueue): "session:{session_id}". Pure — usable by
 * tests and any component validating queue names.
 */
export function formatSessionTaskQueue(sessionId: string): string {
  return SESSION_TASK_QUEUE_PREFIX + sessionId;
}

/**
 * Resolves the activity task queue, harness, and execution target for one
 * execution (Go ResolveActivityTaskQueue). Error message text is pinned:
 * the create step surfaces it verbatim as FailedPrecondition (Go create.go
 * startWorkflowStep).
 */
export async function resolveActivityTaskQueue(
  store: Store,
  sessionId: string,
  config: AgentExecutionTemporalConfig,
  activityTaskQueueOverride: string,
  logger: Logger,
): Promise<DispatchResult> {
  // Sandbox affinity: the parent workflow already has a sandbox on this
  // queue — route there directly, no session routing, no provisioning.
  if (activityTaskQueueOverride !== "") {
    let harness = Harness.NATIVE;
    if (sessionId !== "") {
      try {
        const session = await store.getResource(
          ApiResourceKind.session,
          sessionId,
          SessionSchema,
        );
        // A LOADED session with a nil spec resolves UNSPECIFIED (Go's
        // GetSpec().GetHarness() nil-chain); NATIVE is only the pre-load
        // default.
        harness = session.spec?.harness ?? Harness.UNSPECIFIED;
      } catch {
        // Best-effort read (Go ignores the error and keeps NATIVE).
      }
    }

    logger.info("Dispatch using activity_task_queue override (sandbox affinity)", {
      session_id: sessionId,
      task_queue: activityTaskQueueOverride,
      override_source: "parent_workflow_sandbox",
    });

    return {
      taskQueue: activityTaskQueueOverride,
      harness,
      executionTarget: ExecutionTarget.LOCAL,
    };
  }

  let harness = Harness.NATIVE;
  let executionTarget = ExecutionTarget.UNSPECIFIED;

  if (sessionId !== "") {
    try {
      const session = await store.getResource(
        ApiResourceKind.session,
        sessionId,
        SessionSchema,
      );
      // Loaded-session nil-spec resolves UNSPECIFIED, exactly as the
      // override arm above.
      harness = session.spec?.harness ?? Harness.UNSPECIFIED;
      executionTarget = session.spec?.executionTarget ?? ExecutionTarget.UNSPECIFIED;
    } catch (error) {
      if (!(error instanceof ResourceNotFoundError)) {
        // Pinned text (create maps it to FailedPrecondition verbatim).
        throw new Error(
          `failed to load session for dispatch: ${error instanceof Error ? error.message : String(error)}`,
        );
      }
      logger.warn("Session not found during dispatch, using defaults", {
        session_id: sessionId,
      });
    }
  }

  const resolvedTarget = config.resolveExecutionTarget(executionTarget);
  const taskQueue = resolveTaskQueue(sessionId, config);

  logger.info("Dispatch resolved activity task queue", {
    session_id: sessionId,
    task_queue: taskQueue,
    routing_mode: config.activityRouting,
    execution_target: ExecutionTarget[resolvedTarget],
  });

  return { taskQueue, harness, executionTarget: resolvedTarget };
}

/** Go resolveTaskQueue: session routing only with a session id in hand. */
function resolveTaskQueue(
  sessionId: string,
  config: AgentExecutionTemporalConfig,
): string {
  if (config.activityRouting === ROUTING_SESSION && sessionId !== "") {
    return formatSessionTaskQueue(sessionId);
  }
  return config.runnerQueue;
}
